Add testing for forward over reverse gradgrad

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69956
* #69884
* #69727
* __->__ #69740
* #69690
* #69661

Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

soulitzer committed:
Add testing for forward over reverse gradgrad

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1437,6 +1439,15 @@ def gradgradcheck(
+        "Expected at least one of check_fwd_over_rev or check_rev_over_rev to be True"
+    assert not (check_undefined_grad and not check_rev_over_rev), \
+        "Setting check_undefined_grad=True requires check_rev_over_rev to be True"
+    assert not (check_batched_grad and not check_rev_over_rev), (
```
I'm confused by this one. Today the batched for rev over rev is the one that is running. Why are we disallowing it?

soulitzer replied:
What do you mean by disallowing it? This is the same check we would have when we enter gradcheck, so this one is a bit redundant actually, but this one here will give users a slightly better error.

albanD replied:
Ho actually the check is ok. But the message is weird right?

soulitzer replied:
Oh right, good catch!

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1437,6 +1439,15 @@ def gradgradcheck(
+    assert not (check_batched_grad and not check_rev_over_rev), (
+        "Setting check_batched_grad=True requires check_fwd_over_rev to be True")
+    # TODO: do we want to test this too?
+    # assert not (check_batched_forward_grad and not check_fwd_over_rev), (
```
Is it a no-op otherwise?

soulitzer replied:
Yeah currently there is no such test (since `check_batched_forward_grad` is not passed to gradcheck). The proposal here is to add a new `check_batched_fw_over_rev` flag for `gradgradcheck`, and if `check_batched_fw_over_rev=True`, we pass `check_batched_forward_grad=True` to gradcheck.

albanD replied:
That sounds good!

soulitzer replied:
Saving this for a followup

albanD reviewed (commented):
Nice!

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"




[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"




[ghstack-poisoned]

albanD reviewed (approved):
LGTM

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"




[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"




[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10220,19 +10343,23 @@ def ref_pairwise_distance(input1, input2):
     OpInfo('masked_select',
            dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
            supports_forward_ad=True,
+           # RuntimeError: ZeroTensors are immutable. Please use the materialized zero tensor (...)
```
Can you please open an issue for this so we can track them on the side?

soulitzer replied:
https://github.com/pytorch/pytorch/issues/69855

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031727).

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031727).

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031727).

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031727).

soulitzer committed:
Update on "Add testing for forward over reverse gradgrad"


Differential Revision: [D33031727](https://our.internmc.facebook.com/intern/diff/D33031727)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031727).
